utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,396 @@
+        pos + 2
+    }
+}
+
+/// This struct represents a de-serialized Char16Trie that was exported from
+/// ICU binary data.
+///
+/// The trie consists of a series of char16_t-serialized nodes for incremental
+/// Unicode string/char16_t sequence matching. (char16_t=16-bit unsigned integer)
+/// The root node is at the beginning of the trie data.
+///
+/// Types of nodes are distinguished by their node lead unit ranges.
+/// After each node, except a final-value node, another node follows to
+/// encode match values or continue matching further units.
+///
+/// Node types:
+///  - Final-value node: Stores a 32-bit integer in a compact, variable-length format.
+///    The value is for the string/char16_t sequence so far.
+///  - Match node, optionally with an intermediate value in a different compact format.
+///    The value, if present, is for the string/char16_t sequence so far.
+///
+///  Aside from the value, which uses the node lead unit's high bits:
+///
+///  - Linear-match node: Matches a number of units.
+///  - Branch node: Branches to other nodes according to the current input unit.
+///    The node unit is the length of the branch (number of units to select from)
+///    minus 1. It is followed by a sub-node:
+///    - If the length is at most MAX_BRANCH_LINEAR_SUB_NODE_LENGTH, then
+///      there are length-1 (key, value) pairs and then one more comparison unit.
+///      If one of the key units matches, then the value is either a final value for
+///      the string so far, or a "jump" delta to the next node.
+///      If the last unit matches, then matching continues with the next node.
+///      (Values have the same encoding as final-value nodes.)
+///    - If the length is greater than MAX_BRANCH_LINEAR_SUB_NODE_LENGTH, then
+///      there is one unit and one "jump" delta.
+///      If the input unit is less than the sub-node unit, then "jump" by delta to
+///      the next sub-node which will have a length of length/2.
+///      (The delta has its own compact encoding.)
+///      Otherwise, skip the "jump" delta to the next sub-node
+///      which will have a length of length-length/2.
+///
+/// For more information:
+/// - [ICU4C UCharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UCharsTrie.html)
+/// - [ICU4J CharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/CharsTrie.html) API.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone)]
+pub struct Char16Trie<'data> {
+    /// An array of u16 containing the trie data.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub data: ZeroVec<'data, u16>,
+}
+
+/// This struct represents an iterator over a Char16Trie.
+pub struct Char16TrieIterator<'a> {
+    /// A reference to the Char16Trie data to iterate over.
+    trie: &'a [<u16 as zerovec::ule::AsULE>::ULE],
+    /// Index of next trie unit to read, or None if there are no more matches.
+    pos: Option<usize>,
```

Comment:
FYI -- ICU4J is also using a pos index rather than C++'s pos pointer, and uses a negative pos value for "none". Simpler than `Option<usize>` in Rust, too? (Loads only one int, rather than bool-then-int as in the Option.)

https://github.com/unicode-org/icu/blob/main/icu4j/main/classes/core/src/com/ibm/icu/util/CharsTrie.java#L697

Reply:
Like above, I think Option<> is more idiomatic Rust.

Reply:
That's probably ok for pos, more than for remaining_match_length.